Pass `totalItems` to `prepare_collection_response`

pfefferle opened the pull request:
Refactors the REST collection response preparation to accept an explicit `$total_items` parameter instead of relying on the `totalItems` value from the response array. This improves accuracy in pagination calculations by allowing controllers to pass the correct total item count directly.

## Proposed changes:
- Updated method signature for `prepare_collection_response` to accept `$total_items` parameter
- Modified all controller calls to pass explicit total item counts
- Enhanced outbox totalItems calculation to count published posts more accurately

### Other information:

- [X] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [X] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Added - for new features
-   [X] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Improved REST collection pagination by using explicit total item counts for more accurate results.

</details>

pfefferle committed:
Pass totalItems to prepare_collection_response

Refactors REST controllers to explicitly pass the total number of items to the prepare_collection_response method, improving accuracy in pagination calculations. Updates the method signature in trait-collection.php and all relevant controller calls. Also updates outbox totalItems calculation to better reflect published post counts.

pfefferle commented on includes/rest/class-outbox-controller.php:
```diff
@@ -198,7 +210,7 @@ public function get_items( $request ) {
 		}
 
-		$response = $this->prepare_collection_response( $response, $request );
+		$response = $this->prepare_collection_response( $response, $request, $outbox_query->found_posts );
```
this is on purpose

pfefferle committed:
Refactor prepare_collection_response parameter usage

Updated controllers to call prepare_collection_response without explicitly passing total items, defaulting to use response['totalItems'] if not provided. This simplifies method signatures and centralizes logic for determining total items within the trait.

pfefferle committed:
Fix max_pages calculation in collection response

Replaces null coalescing operator with short ternary for compatibility and disables PHPCS warning. Ensures correct calculation of max_pages when total_items is zero or not provided.

pfefferle committed:
Fix max_pages calculation and remove totalItems test assertions

Corrects the calculation of max_pages in the Collection trait to ensure proper pagination when total_items is provided. Removes several test assertions in Test_Outbox_Controller that directly check the value of totalItems, likely to decouple tests from internal implementation details or to reflect changes in response structure.

pfefferle committed:
Add assertions for content visibility in outbox tests

Additional assertions verify that the number of visible items in the outbox matches expected visibility for both logged-out users and users with the activitypub capability. This improves test coverage for content visibility logic.

matticbot committed:
Add changelog

obenland commented on includes/rest/class-outbox-controller.php:
```diff
@@ -175,10 +175,22 @@ public function get_items( $request ) {
+			$response['totalItems'] = \intval( $count_posts );
+		} else {
+			foreach ( $post_types as $post_type ) {
+				$count_posts             = \wp_count_posts( $post_type );
```
I assume both `count_user_posts()` and `wp_count_posts()` include unfederated posts, not sure how you feel about that?

pfefferle replied:
That was what we used before. I fear that custom queries would make the request unnecessarily slow!?

obenland commented on includes/rest/class-outbox-controller.php:
```diff
@@ -198,7 +210,7 @@ public function get_items( $request ) {
 		}
 
-		$response = $this->prepare_collection_response( $response, $request );
+		$response = $this->prepare_collection_response( $response, $request, $outbox_query->found_posts );
```
Instead of passing `$outbox_query->found_posts` to `prepare_collection_response()`, couldn't we just keep it as the default? `'totalItems'` should be the correct amount before the response gets passed to `prepare_collection_response()`.

pfefferle committed:
Remove deprecated hooks and simplify outbox response

Deprecated filters and actions related to the outbox controller have been removed, and the calculation of totalItems now relies directly on WP_Query's found_posts. The prepare_collection_response method signature was simplified, and related test code for deprecated hooks was also removed.

pfefferle committed:
Update trait-collection.php

pfefferle committed:
Add filter to override outbox totalItems count

Introduces the overload_total_items method and attaches it to the 'activitypub_rest_outbox_array' filter. This sets the 'totalItems' property in the outbox response to 100, ensuring compatibility with Mastodon's expectations for federated post counts.

pfefferle committed:
Calculate totalItems dynamically in outbox response

Update the overload_total_items method to compute the total number of federated posts using WP_Query, based on the user_id and posts with the 'activitypub_status' meta key. This replaces the previous hardcoded value and ensures accurate counts for clients like Mastodon.

pfefferle committed:
Include comments in outbox totalItems count

Updated the outbox controller to add the number of approved comments with the 'activitypub_status' meta key to the totalItems property, ensuring Mastodon and other clients see the combined count of federated posts and comments.

pfefferle committed:
Merge branch 'trunk' into fix/post-count

obenland commented on includes/rest/class-outbox-controller.php:
```diff
@@ -273,4 +271,49 @@ public function get_item_schema() {
+					),
+				),
+				'fields'      => 'ids',
+				'nopaging'    => true,
```
Since we don't do anything with these IDs, I wonder if it would be a faster query if we just looked for 1 post, similar to what we're doing with comments below?

pfefferle committed:
Update WP_Query args in overload_total_items method

Replaces 'nopaging' with 'no_found_rows' and adds 'number' parameter set to 1 in the WP_Query arguments. This may improve query performance and ensure correct total item calculation.
